### ISSUE TITLE
Formally define OHDSI-SQL

### ISSUE THREAD

**Issue by ablack3** [OHDSI/SqlRender#290]
SqlRender performs substitution on strings of text but does not guarantee that it's output will actually run on the desired dbms. This is because it does not check that it's input is valid OHDSI-SQL.

What if we thought of OHDSI-SQL as a sql dialect and formally defined it's structure so that we could

- Check that inputs are valid and can be translated
- Know that we have covered all translation cases when adding a new backend
- Potentially provide editor support for code completion and syntax checking

The definition would look something like
<img width="855" alt="image" src="https://user-images.githubusercontent.com/10227522/183087379-d07fd3e4-184e-437a-bfc3-f87274ec241a.png">

**Comment by schuemie** [comment]
This is somewhat related to https://github.com/OHDSI/SqlRender/issues/157, but I like the idea of being able to validate the input. I looked into SQL parsers in the past, but found them hard to implement in the SqlRender syntax (i.e. I couldn't find a good Java library).

**Comment by ablack3** [comment]
Yea I think this is really the same issue as #157 but with a focus on a formal definition of OHDSI-SQL syntax that could be checked and parsed programmatically possibly with [javacc](https://javacc.github.io/javacc/). 

Artem implemented duckdb sql translations but we don't really know if we have implemented all of OHDSI SQL because we don't really know the full domain of inputs to `translate`.

**Comment by schuemie** [comment]
Yes, I tend to use [this section](https://ohdsi.github.io/SqlRender/articles/UsingSqlRender.html#functions-and-structures-supported-by-translate) as an informal definition of the scope of SqlRender.

**Comment by ablack3** [comment]
@SofiaMp from Kheiron is interested in taking this on. My description of this task is that we want to formally specify the OHDSI-SQL language since it is a programming language.

I would also suggest we separately specify the inputs to `render` and `translate`.


render: {parameterized OHDSI SQL} -> {non parameterized OHDSI SQL}
translate: {non-parameterized OHDSI SQL} -> {sql server, bigquery, impala, ....}

{parameterized OHDSI SQL}  can contain branching logic and parameters while {non-parameterized OHDSI SQL} should not.

We need to define the complete syntax and semantics of OHDSI-SQL. Most of the work will be in defining the syntax. The only semantics that needs to be defined are the non-SQL elements like branching logic, parameters, and default value declarations which are all done in the SqlRender vignettes. We don't need define the semantics of SQL though.

I'll leave it up to @SofiaMp about how to approach this but the result will be a document (a new vignette perhaps) that provides a complete description of the OHDSI-SQL syntax that could be used to implement a [parser](https://cs.lmu.edu/~ray/notes/javacc/).

https://homepage.cs.uri.edu/faculty/hamel/courses/2018/fall2018/csc301/lecture-notes/csc301-ln002.pdf
